Add service monitor for operand

mfojtik opened the pull request.

sttts commented on bindata/v3.11.0/kube-apiserver/prometheus-service-monitor.yaml:
```diff
@@ -0,0 +1,27 @@
+      interval: 30s
+      metricRelabelings:
+        - action: drop
+          regex: etcd_(debugging|disk|request|server).*
```
where does this file come from?

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+	"k8s.io/client-go/dynamic"
+)
+
+var monitorGroupVersionKind = schema.GroupVersionResource{Group: "monitoring.coreos.com", Version: "v1", Resource: "servicemonitors"}
```
monitorGVR, not kind

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+
+var monitorGroupVersionKind = schema.GroupVersionResource{Group: "monitoring.coreos.com", Version: "v1", Resource: "servicemonitors"}
+
+// ApplyServiceMonitor applies the Prometheus service monitor.
```
what does "requires rules" mean?

mfojtik replied:
fixed, bad copy & paste.

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+func ApplyServiceMonitor(client dynamic.Interface, recorder events.Recorder, serviceMonitorBytes []byte) (bool, error) {
+	monitorJSON, err := yaml.YAMLToJSON(serviceMonitorBytes)
+	if err != nil {
+		panic(err)
```
return err

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+	}
+	monitorObj, err := runtime.Decode(unstructured.UnstructuredJSONScheme, monitorJSON)
+	if err != nil {
+		panic(err)
```
return err

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+		panic(fmt.Sprintf("unexpected object in %t", monitorObj))
+	}
+
+	existing, err := client.Resource(monitorGroupVersionKind).Get(required.GetName(), metav1.GetOptions{})
```
no namespace?

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+		panic(fmt.Sprintf("unexpected object in %t", monitorObj))
+	}
+
+	existing, err := client.Resource(serviceMonitorGVR).Get(required.GetName(), metav1.GetOptions{})
```
namespace

sttts commented on pkg/operator/target_config_reconciler_v311_00.go:
```diff
@@ -49,6 +51,10 @@ func createTargetConfigReconciler_v311_00_to_latest(c TargetConfigReconciler, re
 		errors = append(errors, fmt.Errorf("%q: %v", "configmap/kube-apiserver-pod", err))
 	}
+	_, err = managePrometheus_v311_00_to_latest(c.kubeClient.RbacV1(), c.dynamicClient, recorder)
+	if err != nil {
```
`if _, err := ... ; err != nil `

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+	"fmt"
+
+	"github.com/ghodss/yaml"
+	"github.com/openshift/library-go/pkg/operator/events"
```
order

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+		_, createErr := client.Resource(serviceMonitorGVR).Create(required)
+		if createErr != nil {
+			recorder.Warningf("ServiceMonitorCreateFailed", "Failed to create ServiceMonitor.monitoring.coreos.com/v1: %v", createErr)
+			return true, createErr
```
not `return false, createErr`?

mfojtik replied:
no, we return true in this case from other apply func... i guess it indicates that we detected a change (not found) but failed to react to it (error)

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+		return true, err
+	}
+
+	_, err = client.Resource(serviceMonitorGVR).Update(existing)
```
namespace

if _, err := ...; err != nil

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	_, err = client.Resource(serviceMonitorGVR).Update(existing)
+	if err == nil {
```
!=

sttts commented on vendor/github.com/openshift/library-go/pkg/operator/resource/resourceapply/prometheus.go:
```diff
@@ -0,0 +1,71 @@
+		recorder.Eventf("ServiceMonitorUpdated", "Updated ServiceMonitor.monitoring.coreos.com/v1 because it changed")
+	}
+
+	return true, err
```
true only in `err == nil` case

mfojtik replied:
i think same logic as create applies here (we found a change, we failed to react (so true, error))

mfojtik commented:
@sttts thanks for review! addressed all comments except 2 ;-)

@deads2k if you're happy with this (and tests will pass), I will do real bump in library-go and make similar change in openshift-apiserver.

I'm not sure about controller manager, I think we postponed serving metrics/etc. as we can't serve them securely until we get that in 1.12 rebase?

mfojtik commented:
/retest

deads2k commented on bindata/v3.11.0/kube-apiserver/svc.yaml:
```diff
@@ -5,12 +5,10 @@ metadata:
 spec:
   type: ClusterIP
   selector:
-    apiserver: "true"
```
> do we use this label for anything? if not I would drop it from the pod maybe?

sure

deads2k commented on pkg/operator/target_config_reconciler_v311_00.go:
```diff
@@ -50,6 +52,10 @@ func createTargetConfigReconciler_v311_00_to_latest(c TargetConfigReconciler, re
 		errors = append(errors, fmt.Errorf("%q: %v", "configmap/kube-apiserver-pod", err))
 	}
 
+	if _, err = managePrometheus_v311_00_to_latest(c.kubeClient.RbacV1(), c.dynamicClient, recorder); err != nil {
```
model all this as a separate control loop.  This loop shouldn't trigger a failing status and is logically generic for all of our operators.

mfojtik replied:
i think it makes sense to make this generic. https://github.com/openshift/library-go/pull/150

mfojtik commented:
replaced by https://github.com/openshift/library-go/pull/150
